File: latlon-ellipsoidal.vb
Imports System

' Vector3d class
Public Class Vector3d
    Public Property x As Double
    Public Property y As Double
    Public Property z As Double

    Public Sub New(x As Double, y As Double, z As Double)
        Me.x = x
        Me.y = y
        Me.z = z
    End Sub

    Public Function Length() As Double
        Return Math.Sqrt(x * x + y * y + z * z)
    End Function

    Public Overrides Function ToString() As String
        Return $"({x}, {y}, {z})"
    End Function
End Class

' Dms class
Public Class Dms
    Public Shared Function Wrap360(deg As Double) As Double
        Return If(deg >= 0, deg Mod 360, 360 + (deg Mod 360))
    End Function

    Public Shared Function ToDegrees(radians As Double) As Double
        Return radians * 180.0 / Math.PI
    End Function

    Public Shared Function ToRadians(degrees As Double) As Double
        Return degrees * Math.PI / 180.0
    End Function

    Public Shared Function ToDms(deg As Double) As String
        Dim d As Integer = CInt(Math.Floor(deg))
        Dim m As Integer = CInt(Math.Floor((deg - d) * 60))
        Dim s As Double = ((deg - d) * 60 - m) * 60
        Return $"{d}°{m}'{s}"""
    End Function

    Public Shared Function ParseDms(dms As String) As Double
        Dim parts = dms.Split("°"c, "'"c, """"c)
        If parts.Length <> 3 Then
            Throw New FormatException("Invalid DMS format.")
        End If

        Dim deg As Double = Convert.ToDouble(parts(0))
        Dim min As Double = Convert.ToDouble(parts(1))
        Dim sec As Double = Convert.ToDouble(parts(2))

        Return deg + min / 60 + sec / 3600
    End Function
End Class

' LatLonEllipsoidal class
Public Class LatLonEllipsoidal
    Private _lat As Double
    Private _lon As Double
    Private _height As Double

    Public Property Lat As Double
        Get
            Return _lat
        End Get
        Set(value As Double)
            _lat = Dms.Wrap360(value)
        End Set
    End Property

    Public Property Lon As Double
        Get
            Return _lon
        End Get
        Set(value As Double)
            _lon = Dms.Wrap180(value)
        End Set
    End Property

    Public Property Height As Double
        Get
            Return _height
        End Get
        Set(value As Double)
            _height = value
        End Set
    End Property

    Public Sub New(lat As Double, lon As Double, Optional height As Double = 0)
        _lat = Dms.Wrap360(lat)
        _lon = Dms.Wrap180(lon)
        _height = height
    End Sub

    Public Overrides Function ToString() As String
        Return $"Lat: {_lat}, Lon: {_lon}, Height: {_height}"
    End Function
End Class

' Cartesian class
Public Class Cartesian
    Inherits Vector3d

    Public Sub New(x As Double, y As Double, z As Double)
        MyBase.New(x, y, z)
    End Sub

    Public Function ToLatLon(ellipsoid As Ellipsoid) As LatLonEllipsoidal
        If ellipsoid Is Nothing Then
            ellipsoid = Ellipsoids.WGS84
        End If

        Dim a As Double = ellipsoid.A
        Dim b As Double = ellipsoid.B
        Dim f As Double = ellipsoid.F

        Dim e2 As Double = 2 * f - f * f
        Dim ε2 As Double = e2 / (1 - e2)
        Dim p As Double = Math.Sqrt(x * x + y * y)
        Dim R As Double = Math.Sqrt(p * p + z * z)

        Dim tanβ As Double = (b * z) / (a * p) * (1 + ε2 * b / R)
        Dim sinβ As Double = tanβ / Math.Sqrt(1 + tanβ * tanβ)
        Dim cosβ As Double = sinβ / tanβ

        Dim φ As Double
        If Not Double.IsNaN(cosβ) Then
            φ = Math.Atan2(z + ε2 * b * sinβ * sinβ * sinβ, p - e2 * a * cosβ * cosβ * cosβ)
        Else
            φ = 0
        End If

        Dim λ As Double = Math.Atan2(y, x)

        Dim sinφ As Double = Math.Sin(φ)
        Dim cosφ As Double = Math.Cos(φ)
        Dim ν As Double = a / Math.Sqrt(1 - e2 * sinφ * sinφ)
        Dim h As Double = p * cosφ + z * sinφ - (a * a / ν)

        Return New LatLonEllipsoidal(Dms.ToDegrees(φ), Dms.ToDegrees(λ), h)
    End Function

    Public Overrides Function ToString() As String
        Return $"Cartesian: {MyBase.ToString()}"
    End Function
End Class

' Ellipsoid class
Public Class Ellipsoid
    Public Property A As Double
    Public Property B As Double
    Public Property F As Double

    Public Sub New(a As Double, b As Double, f As Double)
        Me.A = a
        Me.B = b
        Me.F = f
    End Sub
End Class

' Ellipsoids class
Public Class Ellipsoids
    Public Shared ReadOnly WGS84 As New Ellipsoid(6378137, 6356752.314245, 1 / 298.257223563)
End Class

File: latlon-ellipsoidal.js
import Dms from './dms.js';
import Vector3d from './vector3d.js';

//ellipsoid parametes; exposted through static getter below - only ellipsoid defined in WGS84, for use in utm/mgrs, vincenty, nvector
const ellipsoids = {
    WGS84: { a: 6378137, b: 6356752.314245, f: 1 / 298.257223563 },
};

//datums; exposed through static getter below - only datum defined is WGS84, for use in utm/mgrs,vincenty, nvector.
const datums = {
    WGS84: { ellipsoid: ellipsoids.WGS84 },
};

// freeze static properties
Object.freeze(ellipsoids.WGS84);
Object.freeze(datums.WGS84);



/* LatLon Ellipsoidal - - - - - - - - - - - - - - - - - - - - - - - - */

/*
 * Ellisoid parameters and methods used for converting points to/from cartesian coordinates
 *
 * 
 * 
 * This is the core class which will (usually) be used via LatLonEllipsoidal_Datum or LatLonEllipsoidal_ReferenceFrame.
 */

class LatLonEllipsoidal {

    //creates a geodetic lat/lon point on a WGS84 ellipsoidal model earth
    constructor(lat, lon, height = 0) {
        if (isNaN(lat) || lat == null) throw new TypeError(`invalid lat ‘${lat}’`);
        if (isNaN(lon) || lon == null) throw new TypeError(`invalid lon ‘${lon}’`);
        if (isNaN(height) || height == null) throw new TypeError(`invalid height ‘${height}’`);

        this._lat = Dms.wrap90(Number(lat));
        this._lon = Dms.wrap180(Number(lon));
        this._height = Number(height);
    }

    //latitude in degree north from equator
    get lat() { return this._lat; }
    get latitude() { return this._lat; }
    set lat(lat) {
        this._lat = isNaN(lat) ? Dms.wrap90(Dms.parse(lat)) : Dms.wrap90(Number(lat));
        if (isNaN(this._lat)) throw new TypeError(`invalid lat ‘${lat}’`);
    }
    set latitude(lat) {
        this._lat = isNaN(lat) ? Dms.wrap90(Dms.parse(lat)) : Dms.wrap90(Number(lat));
        if (isNaN(this._lat)) throw new TypeError(`invalid latitude ‘${lat}’`);
    }

    //longitude in degrees east from international reference meridian
    get lon() { return this._lon; }
    get lng() { return this._lon; }
    get longitude() { return this._lon; }
    set lon(lon) {
        this._lon = isNaN(lon) ? Dms.wrap180(Dms.parse(lon)) : Dms.wrap180(Number(lon));
        if (isNaN(this._lon)) throw new TypeError(`invalid lon ‘${lon}’`);
    }
    set lng(lon) {
        this._lon = isNaN(lon) ? Dms.wrap180(Dms.parse(lon)) : Dms.wrap180(Number(lon));
        if (isNaN(this._lon)) throw new TypeError(`invalid lng ‘${lon}’`);
    }
    set longitude(lon) {
        this._lon = isNaN(lon) ? Dms.wrap180(Dms.parse(lon)) : Dms.wrap180(Number(lon));
        if (isNaN(this._lon)) throw new TypeError(`invalid longitude ‘${lon}’`);
    }

    //Height in meters above ellipsoid
    get height() { return this._height; }
    set height(height) { this._height = Number(height); if (isNaN(this._height)) throw new TypeError(`invalid height ‘${height}’`); }


    //Datum. NOTE: replicated w/in LatLonEllipsoidal for LatLonEllipsoidal object to be monkey-patched to look like a LatLonEllipsoidal_Datum - for Vincenty calculations on different ellipsoids.
    get datum() { return this._datum; }
    set datum(datum) { this._datum = datum; }


    //Ellipsoid w/their paremeters, this module only defines WGS84 parameters
    static get ellipsoids() {
        return ellipsoids;
    }

    //Datums. -module only defines WGS84 datum
    static get datums() {
        return datums;
    }

    //parses a latitude/longitude point form a variety of formats
    static parse(...args) {
        if (args.length == 0) throw new TypeError('invalid (empty) point');

        let lat = undefined, lon = undefined, height = undefined;

        // single { lat, lon } object
        if (typeof args[0] == 'object' && (args.length == 1 || !isNaN(parseFloat(args[1])))) {
            const ll = args[0];
            if (ll.type == 'Point' && Array.isArray(ll.coordinates)) { // GeoJSON
                [lon, lat, height] = ll.coordinates;
                height = height || 0;
            } else { // regular { lat, lon } object
                if (ll.latitude != undefined) lat = ll.latitude;
                if (ll.lat != undefined) lat = ll.lat;
                if (ll.longitude != undefined) lon = ll.longitude;
                if (ll.lng != undefined) lon = ll.lng;
                if (ll.lon != undefined) lon = ll.lon;
                if (ll.height != undefined) height = ll.height;
                lat = Dms.wrap90(Dms.parse(lat));
                lon = Dms.wrap180(Dms.parse(lon));
            }
            if (args[1] != undefined) height = args[1];
            if (isNaN(lat) || isNaN(lon)) throw new TypeError(`invalid point ‘${JSON.stringify(args[0])}’`);
        }

        // single comma-separated lat/lon
        if (typeof args[0] == 'string' && args[0].split(',').length == 2) {
            [lat, lon] = args[0].split(',');
            lat = Dms.wrap90(Dms.parse(lat));
            lon = Dms.wrap180(Dms.parse(lon));
            height = args[1] || 0;
            if (isNaN(lat) || isNaN(lon)) throw new TypeError(`invalid point ‘${args[0]}’`);
        }

        // regular (lat, lon) arguments
        if (lat == undefined && lon == undefined) {
            [lat, lon] = args;
            lat = Dms.wrap90(Dms.parse(lat));
            lon = Dms.wrap180(Dms.parse(lon));
            height = args[2] || 0;
            if (isNaN(lat) || isNaN(lon)) throw new TypeError(`invalid point ‘${args.toString()}’`);
        }

        return new this(lat, lon, height); // 'new this' as may return subclassed types
    }

    //converts 'this' point from latitude/longitude coordinates to cartesian coordinates
    toCartesian() {
        // x = (ν+h)⋅cosφ⋅cosλ, y = (ν+h)⋅cosφ⋅sinλ, z = (ν⋅(1-e²)+h)⋅sinφ
        // where ν = a/√(1−e²⋅sinφ⋅sinφ), e² = (a²-b²)/a² or (better conditioned) 2⋅f-f²
        const ellipsoid = this.datum
            ? this.datum.ellipsoid
            : this.referenceFrame ? this.referenceFrame.ellipsoid : ellipsoids.WGS84;

        const φ = this.lat.toRadians();
        const λ = this.lon.toRadians();
        const h = this.height;
        const { a, f } = ellipsoid;

        const sinφ = Math.sin(φ), cosφ = Math.cos(φ);
        const sinλ = Math.sin(λ), cosλ = Math.cos(λ);

        const eSq = 2 * f - f * f;                      // 1st eccentricity squared ≡ (a²-b²)/a²
        const ν = a / Math.sqrt(1 - eSq * sinφ * sinφ); // radius of curvature in prime vertical

        const x = (ν + h) * cosφ * cosλ;
        const y = (ν + h) * cosφ * sinλ;
        const z = (ν * (1 - eSq) + h) * sinφ;

        return new Cartesian(x, y, z);
    }

    //checks if another point is equal to 'this' point. 
    equals(point) {
        if (!(point instanceof LatLonEllipsoidal)) throw new TypeError(`invalid point ‘${point}’`);

        if (Math.abs(this.lat - point.lat) > Number.EPSILON) return false;
        if (Math.abs(this.lon - point.lon) > Number.EPSILON) return false;
        if (Math.abs(this.height - point.height) > Number.EPSILON) return false;
        if (this.datum != point.datum) return false;
        if (this.referenceFrame != point.referenceFrame) return false;
        if (this.epoch != point.epoch) return false;

        return true;
    }

        //return a string representation of 'this' point, formatted as degrees, degrees+minutes, or degrees+minutes+seconds
    toString(format = 'd', dp = undefined, dpHeight = null) {
        // note: explicitly set dp to undefined for passing through to toLat/toLon
        if (!['d', 'dm', 'dms', 'n'].includes(format)) throw new RangeError(`invalid format ‘${format}’`);

        const height = (this.height >= 0 ? ' +' : ' ') + this.height.toFixed(dpHeight) + 'm';
        if (format == 'n') { // signed numeric degrees
            if (dp == undefined) dp = 4;
            const lat = this.lat.toFixed(dp);
            const lon = this.lon.toFixed(dp);
            return `${lat}, ${lon}${dpHeight == null ? '' : height}`;
        }

        const lat = Dms.toLat(this.lat, format, dp);
        const lon = Dms.toLon(this.lon, format, dp);

        return `${lat}, ${lon}${dpHeight == null ? '' : height}`;
    }

}

/* CARTESIAN - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

 //ECEF geocentric cartesian coordinates

class Cartesian extends Vector3d {
    constructor(x, y, z) {
        super(x, y, z); //specifies units & axes
    }

    toLatLon(ellipsoid = ellipsoids.WGS84) {
        //NOTE: ellipsoid is available as a parameter for when toLatLon gets subclassed to Ellipsoidal_Datum/Ellipsoidal_Referenceframe.
        if (!ellipsoid || !ellipsoid.a) throw new TypeError(`invalid ellipsoid ‘${ellipsoid}’`);

        const { x, y, z } = this;
        const { a, b, f } = ellipsoid;

        const e2 = 2 * f - f * f;           // 1st eccentricity squared ≡ (a²−b²)/a²
        const ε2 = e2 / (1 - e2);         // 2nd eccentricity squared ≡ (a²−b²)/b²
        const p = Math.sqrt(x * x + y * y); // distance from minor axis
        const R = Math.sqrt(p * p + z * z); // polar radius

        // parametric latitude (Bowring eqn.17, replacing tanβ = z·a / p·b)
        const tanβ = (b * z) / (a * p) * (1 + ε2 * b / R);
        const sinβ = tanβ / Math.sqrt(1 + tanβ * tanβ);
        const cosβ = sinβ / tanβ;

        // geodetic latitude (Bowring eqn.18: tanφ = z+ε²⋅b⋅sin³β / p−e²⋅cos³β)
        const φ = isNaN(cosβ) ? 0 : Math.atan2(z + ε2 * b * sinβ * sinβ * sinβ, p - e2 * a * cosβ * cosβ * cosβ);

        // longitude
        const λ = Math.atan2(y, x);

        // height above ellipsoid (Bowring eqn.7)
        const sinφ = Math.sin(φ), cosφ = Math.cos(φ);
        const ν = a / Math.sqrt(1 - e2 * sinφ * sinφ); // length of the normal terminated by the minor axis
        const h = p * cosφ + z * sinφ - (a * a / ν);

        const point = new LatLonEllipsoidal(φ.toDegrees(), λ.toDegrees(), h);

        return point;
    }

    toString(dp = 0) {
        const x = this.x.toFixed(dp), y = this.y.toFixed(dp), z = this.z.toFixed(dp);
        return `[${x},${y},${z}]`;
    }

}

/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */
export { LatLonEllipsoidal as default, Cartesian, Vector3d, Dms };